weeks/week04/week04.md

```diff
@@ -1,3 +1,482 @@
 # Week 4: Introduction to Object-Oriented Programming (OOP) for Test Automation
 
-Coming soon.
+We will cover how to apply Object-Oriented Programming (OOP) concepts in `Python` to write scalable and maintainable test scripts. OOP helps create structured test automation frameworks where code is modular, reusable, and easy to manage.
+
+Why OOP matters in Test Automation?
+- **Modularity:** Test scripts become easier to maintain when they are modularized using classes.
+- **Reusability:** Common functionalities can be abstracted into base classes and reused across different test cases.
+- **Scalability:** OOP allows test frameworks to grow by organizing test cases and methods into logical units.
+
+
+## References
+- [Python Classes and Objects - Official Documentation](https://docs.python.org/3/tutorial/classes.html)
+- [Object-Oriented Programming in Python - Real Python](https://realpython.com/python3-object-oriented-programming/)
+- [Inheritance in Python - Programiz](https://www.programiz.com/python-programming/inheritance)
+- [Understanding MRO (Method Resolution Order) for old and new style classes in python](https://medium.com/geekculture/understanding-mro-method-resolution-order-for-old-and-new-style-classes-in-python-f541747c48c7)
+
+---
+
+
+## 4.1 OOP Basics
+
+
+## 4.1.1 Classes and Objects
+
+A `class` is a blueprint (like a plan) for creating objects. It defines attributes (data or state) and methods (behavior) for objects. We can understand a `class` as an abstraction from the real world.
+
+An `object` is an instance of a `class`. Each `object` created from a `class` can have unique attributes. It is a concrete class representation.
```

Comment:
The statement _'Each `object` created from a `class` can have unique attributes.'_
1 looks weird as English statement:
should be like 
'Each `object` **is** created from a `class` **and** can have unique attributes. 
2 From statement is not clear:
2.1 Can be 'object' to be created not from class?
2.2 Can 'object have not unique attributes? If yes, statement  has no sense; if not we should you 'must' instead of 'can'.

Reply:
Fixed English statements for better clarity. Regarding the other decorators, I added a few words and a reference for the `@abstract...` ones and a block for the `@dataclass`. What do you think?